include/os/linux/kernel/linux/simd_x86.h

```diff
@@ -126,38 +126,64 @@
 #endif
 
 #else /* defined(KERNEL_EXPORTS_X86_FPU) */
+
 /*
  * When the kernel_fpu_* symbols are unavailable then provide our own
  * versions which allow the FPU to be safely used in kernel threads.
  * In practice, this is not a significant restriction for ZFS since the
  * vast majority of SIMD operations are performed by the IO pipeline.
  */
+#if defined(HAVE_KERNEL_FPU_INTERNAL)
 
 /*
- * Returns non-zero if FPU operations are allowed in the current context.
+ * FPU usage only allowed in dedicated kernel threads.
  */
-#if defined(HAVE_KERNEL_TIF_NEED_FPU_LOAD)
-#define	kfpu_allowed()		((current->flags & PF_KTHREAD) && \
-				test_thread_flag(TIF_NEED_FPU_LOAD))
-#elif defined(HAVE_KERNEL_FPU_INITIALIZED)
-#define	kfpu_allowed()		((current->flags & PF_KTHREAD) && \
-				current->thread.fpu.initialized)
-#else
-#define	kfpu_allowed()		0
-#endif
+#define	kfpu_allowed()		(current->flags & PF_KTHREAD)
+#define	ex_handler_fprestore	ex_handler_default
+
+/*
+ * FPU save and restore instructions.
+ */
+#define	__asm			__asm__ __volatile__
+#define	kfpu_fxsave(addr)	__asm("fxsave %0" : "=m" (*(addr)))
+#define	kfpu_fxsaveq(addr)	__asm("fxsaveq %0" : "=m" (*(addr)))
+#define	kfpu_fnsave(addr)	__asm("fnsave %0" : "=m" (*(addr)))
```

Comment:
the kernel assembly has an additional fwait after fnsave, since fsave/fnsave clear the registers. not sure whether it is needed here as well?

Reply:
I can add it, just in case, but I wasn't sure why it would be needed.  I also noticed in the FreeBSD implementation it was not present.